autograd.h

```diff
@@ -382,24 +393,118 @@ AUTOGRAD_OPTIMIZER_CLASS(SGD) {
+  void serialize(Archive & ar) { 
+    ar(CEREAL_NVP(step_buffer_),
+       CEREAL_NVP(exp_avg_buffer_),
+       CEREAL_NVP(exp_avg_sq_buffer_)); 
+  }
   
+ private:
+  friend class cereal::access;
+  Adam() { }
   double lr_;
-  std::unordered_map<std::string, int> step_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_buffer;
-  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer;
+  std::unordered_map<std::string, int> step_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_buffer_;
+  std::unordered_map<std::string, at::Tensor> exp_avg_sq_buffer_;
 };
 
 }  // namespace autograd
+
+// This is super ugly and I don't know how to simplify it
+CEREAL_REGISTER_TYPE(autograd::SGD);
+CEREAL_REGISTER_POLYMORPHIC_RELATION(autograd::OptimizerImpl, autograd::SGD);
+CEREAL_REGISTER_TYPE(autograd::Adam);
+CEREAL_REGISTER_POLYMORPHIC_RELATION(autograd::OptimizerImpl, autograd::Adam);
+
+namespace cereal {
+template<class Archive>
+void save(Archive & archive, at::Tensor const & tensor) { 
+  if (!tensor.defined()) {
+    auto type = at::ScalarType::Undefined;
+    archive(CEREAL_NVP(type));
+    return;
+  } else {
+    auto type = tensor.type().scalarType();
+    archive(CEREAL_NVP(type));
+  }
+  auto sizes = std::vector<int64_t>();
+  auto buf = std::vector<uint8_t>();
+  for (auto s : tensor.sizes()) {
+    sizes.push_back(s);
+  }
+  auto contig = tensor.toBackend(at::kCPU).contiguous();
+  auto size = tensor.storage()->size() * tensor.storage()->elementSize();
+  at::Backend backend = tensor.type().backend();
+
+  buf.resize(size);
+  memcpy(buf.data(), contig.storage()->data(), size);
+
+  archive(CEREAL_NVP(backend), CEREAL_NVP(sizes), CEREAL_NVP(buf)); 
+}
+
+/**
+ * We follow these rules for loading:
+ * 1. If tensor is defined, and the same ScalarType as the saved tensor,
+ *    then we simply copy the data into the tensor, with resizing.
+ * 2. Otherwise, overwrite the provided tensor with the right type and backend
+ **/
+template<class Archive>
+void load(Archive & archive, at::Tensor & tensor) {
+  at::ScalarType type;
+  archive(CEREAL_NVP(type));
+  if (type == at::ScalarType::Undefined) {
+    tensor = at::Tensor();
+    return;
+  }
+
+  at::Backend backend;
+  auto sizes = std::vector<int64_t>();
+  auto buf = std::vector<uint8_t>();
+  archive(CEREAL_NVP(backend), CEREAL_NVP(sizes), CEREAL_NVP(buf)); 
+
+  if (!tensor.defined() || tensor.type().scalarType() != type) {
+    tensor = at::getType(backend, type).tensor();
+  }
+  if (tensor.type().is_cuda()) {
+    // should actually use cudamemcpy probably
+    auto cputensor = at::CPU(tensor.type().scalarType()).tensor(sizes);
+    tensor.resize_(sizes);
+    memcpy(cputensor.storage()->data(), buf.data(), buf.size());
+    tensor.copy_(cputensor);
+  } else {
+    tensor.resize_(sizes);
+    memcpy(tensor.storage()->data(), buf.data(), buf.size());
+  }
+} 
+
+template<class Archive>
+void load(Archive & archive, tag::Variable & tensor) {
```

Comment:
`save()` works out of the box for variables? And maybe rename the parameter to var.

Reply:
It should, though it won't save gradients.